### PR TITLE
[MUGEN] Add Video VQVAE encoder and decoder

### PR DESCRIPTION
**Pull request by RdoubleA**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #101
* #98
* __->__ #87
* #86

Add Video VQVAE specific encoder and decoder from [MUGEN](https://github.com/mugen-org/MUGEN_baseline/blob/main/lib/models/video_vqvae/vqvae.py).

Differential Revision: [D37288372](https://our.internmc.facebook.com/intern/diff/D37288372)

### CODE REVIEW

File: torchmultimodal/models/video_vqvae.py [hunk 1]
```diff
@@ -4,10 +4,162 @@
...
+            nn.ReLU(),
+        )
+        self.convts = nn.ModuleList(
+            [
+                SamePadConvTranspose3d(i, o, k, s, bias=True, **kwargs)
+                for i, o, k, s in zip(in_channels, out_channels, kernel_sizes, strides)
+            ]
+        )
+
+    def forward(self, x: Tensor):
```
Comment (ebsmothers) [inline torchmultimodal/models/video_vqvae.py:150]: nit: add return type

File: torchmultimodal/models/video_vqvae.py [hunk 2]
```diff
@@ -4,10 +4,162 @@
...
+    ):
+        super().__init__()
+        if not (
+            len(in_channels) == len(out_channels) == len(kernel_sizes) == len(strides)
+        ):
+            raise ValueError(
+                "in_channels, out_channels, kernel_sizes, strides should all have the same length"
+            )
+        if in_channels[1:] != out_channels[:-1]:
+            raise ValueError("out_channels should match in_channels offset by one")
```
Comment (ebsmothers) [inline torchmultimodal/models/video_vqvae.py:60]: I find we are doing these checks a lot for conv layers. Maybe we should write a util for it; I guess the tradeoff is that it may come at the cost of code readability. Wdyt?

Comment (RdoubleA) [reply torchmultimodal/models/video_vqvae.py:60]: yup agree, was thinking of doing this in a separate PR

File: torchmultimodal/models/video_vqvae.py [hunk 3]
```diff
@@ -4,10 +4,161 @@
...
+class VideoEncoder(nn.Module):
+    """Encoder for Video VQVAE. Stacks specified number of ``SamePadConv3d`` layers
+    followed by a stack of ``AttentionResidualBlocks`` and a final ``SamePadConv3d``
+    layer before the codebook. The residual blocks use Axial Attention to enhance
+    representations of video data without significantly increasing computational
+    cost. Follows VideoGPT's implementation:
+    https://github.com/wilson1yan/VideoGPT/blob/master/videogpt/vqvae.py
+
+    Args:
+        in_channels (Tuple[int]): tuple of input channel dimension for each conv layer
```
Comment (ankitade) [inline torchmultimodal/models/video_vqvae.py:24]: call it in_channels_dim? same for out

File: torchmultimodal/models/video_vqvae.py [hunk 4]
```diff
@@ -4,10 +4,161 @@
...
+        out_channels (Tuple[int]): tuple of output channel dimension for each conv layer
+        kernel_sizes (Tuple[int or Tuple[int]]): tuple of kernel sizes for each conv layer
+        strides (Tuple[int or Tuple[int]]): tuple of strides for each conv layer
+        n_res_layers (int): number of ``AttentionResidualBlocks`` to include
+        embedding_dim (int): size of hidden dimension of final output
+        **kwargs (dict): keyword arguments to be passed into ``SamePadConv3d`` and used by ``nn.Conv3d``
```
Comment (ankitade) [inline torchmultimodal/models/video_vqvae.py:30]: hm do we really need kwargs vs making it explicit?

File: torchmultimodal/models/video_vqvae.py [hunk 5]
```diff
@@ -4,10 +4,161 @@
...
+        if not (
+            len(in_channels) == len(out_channels) == len(kernel_sizes) == len(strides)
+        ):
+            raise ValueError(
+                "in_channels, out_channels, kernel_sizes, strides should all have the same length"
+            )
+        if in_channels[1:] != out_channels[:-1]:
+            raise ValueError("out_channels should match in_channels offset by one")
+
+        convs: List[nn.Module] = []
```
Comment (ankitade) [inline torchmultimodal/models/video_vqvae.py:61]:
```suggestion
        convolutions: List[nn.Module] = []
```

File: test/models/test_video_vqvae.py [hunk 6]
```diff
@@ -47,3 +67,105 @@ def test_forward(self, random):
...
+                    [
+                        [[0.4504, 0.2898], [0.2898, 0.2811]],
+                        [[0.3761, 0.0409], [0.5910, 0.4858]],
+                    ],
+                ]
+            ]
+        )
+        assert_expected(actual, expected, rtol=0, atol=1e-4)
+
+    def test_res_stack_has_attn(self, encoder):
```
Comment (ankitade) [inline test/models/test_video_vqvae.py:110]: do we need separate tests for this and below?

File: torchmultimodal/models/video_vqvae.py (hunk 3 above)
Comment (RdoubleA) [reply torchmultimodal/models/video_vqvae.py:24]: I'm hesitant to change that, since I've kept this consistent across all components all the way down to the conv layers

File: torchmultimodal/models/video_vqvae.py (hunk 4 above)
Comment (RdoubleA) [reply torchmultimodal/models/video_vqvae.py:30]: kwargs are fed into `nn.Conv3d`. I'm not sure if there are any arguments for `nn.Conv3d` the user might want to have control of at this level. maybe `device`, which I could bring out explicitly, but then I should also bring that out in `SamePadConv3d`. `padding_mode` shouldn't be handled at this level. `dilation` I have not tested with the same padding. maybe I should just remove kwargs entirely. wdyt?

File: test/models/test_video_vqvae.py (hunk 6 above)
Comment (RdoubleA) [reply test/models/test_video_vqvae.py:110]: I'll delete `res_stack_has_attn` but I'd like to keep the other to test that the structure is correct

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37288372).

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37288372).

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37288372).

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37288372).

File: torchmultimodal/utils/assertion.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Sized
+
+
+def assert_equal_lengths(*args: Sized, msg="iterable arguments must have same length."):
```
Comment (langong347) [inline torchmultimodal/utils/assertion.py:10]: Why cannot fit this into `common.py`? Or just a first step to split the ever growing `common.py`?

**Review by langong347 — APPROVED** [review]

File: torchmultimodal/utils/assertion.py (hunk 7 above)
Comment (RdoubleA) [reply torchmultimodal/utils/assertion.py:10]: yes, common was getting crowded and I think error checking utilities should have their own files

**Comment by RdoubleA** [comment]
@RdoubleA has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37288372).
